src/NuGet.Clients/NuGet.PackageManagement.UI/Options/PackageSourceMappingOptionsControl.xaml.cs

```diff
@@ -0,0 +1,217 @@
+            return true;
+        }
+
+        // Returns true if there are changes between existingSourceMappings and packageSourceMappings.
+        private static bool SourceMappingsChanged(IReadOnlyList<PackageSourceMappingSourceItem> existingSourceMappings, IReadOnlyList<PackageSourceMappingSourceItem> packageSourceMappings)
+        {
+            if (existingSourceMappings.Count != packageSourceMappings.Count)
+            {
+                return true;
+            }
+
+            for (int i = 0; i < existingSourceMappings.Count; ++i)
+            {
+                //checks to see if keys match
+                if (existingSourceMappings[i] != packageSourceMappings[i])
+                {
+                    return true;
+                }
+                //checks to see if all patterns match
+                if (existingSourceMappings[i].Patterns.Count != packageSourceMappings[i].Patterns.Count)
+                {
+                    return true;
+                }
+                for (int j = 0; j < existingSourceMappings[i].Patterns.Count; ++j)
+                {
+                    if (existingSourceMappings[i].Patterns[j] != packageSourceMappings[i].Patterns[j])
+                    {
+                        return true;
+                    }
+                }
+            }
+            return false;
+        }
+
+        private IReadOnlyList<SourceMappingViewModel> CreateViewModels(IReadOnlyList<PackageSourceMappingSourceItem> originalMappings)
+        {
+            var uiSourceMappings = new Dictionary<string, List<PackageSourceContextInfo>>();
+            foreach (PackageSourceMappingSourceItem sourceItem in originalMappings)
+            {
+                foreach (PackagePatternItem patternItem in sourceItem.Patterns)
+                {
+                    if (!uiSourceMappings.ContainsKey(patternItem.Pattern))
+                    {
+                        uiSourceMappings[patternItem.Pattern] = new List<PackageSourceContextInfo>();
+                    }
+                    uiSourceMappings[patternItem.Pattern].Add(new PackageSourceContextInfo(sourceItem.Key));
+                }
+            }
+            var mappingsCollection = new List<SourceMappingViewModel>();
+            foreach (string packageID in uiSourceMappings.Keys)
+            {
+                var viewModel = new SourceMappingViewModel(packageID, uiSourceMappings[packageID]);
+                mappingsCollection.Add(viewModel);
+            }
+            mappingsCollection.Sort((a, b) => StringComparer.OrdinalIgnoreCase.Compare(a.ID, b.ID));
+            return mappingsCollection.AsReadOnly();
+        }
+
+        private IReadOnlyList<PackageSourceMappingSourceItem> ConvertViewModelToSourceMappingsSourceItems(ItemsChangeObservableCollection<SourceMappingViewModel> sourceMappingViewModels)
+        {
+            Dictionary<string, List<PackagePatternItem>> mappingsDictonary = new Dictionary<string, List<PackagePatternItem>>();
```

Comment:
Just spotted this one. 
```suggestion
            var mappingsDictionary = new Dictionary<string, List<PackagePatternItem>>();
```

Also, what's a better name? Maybe `sourceNamesToPackagePatterns` ??